src/inmanta/compiler/help/explainer.py

```diff
@@ -0,0 +1,166 @@
+
+        if not explainable:
+            return []
+        else:
+            return [self.do_explain(x) for x in explainable]
+
+    def do_explain(self, problem: CompilerException) -> str:
+        env = Environment(
+            loader=PackageLoader('inmanta.compiler.help'),
+        )
+        for name, filter in CUSTOM_FILTERS.items():
+            env.filters[name] = filter
+
+        template = env.get_template(self.template)
+        return template.render(**self.get_arguments(problem))
+
+    def get_arguments(self, problem: CompilerException) -> Dict[str, Any]:
+        return {}
+
+
+class ModifiedAfterFreezeExplainer(JinjaExplainer):
+
+    def __init__(self):
+        super(ModifiedAfterFreezeExplainer, self).__init__("modified_after_freeze.j2", ModifiedAfterFreezeException)
+
+    def build_reverse_hint(self, problem):
+        if isinstance(problem.stmt, AssignStatement):
+            return "%s.%s = %s" % (problem.stmt.rhs.pretty_print(),
+                                   problem.attribute.get_name(),
+                                   problem.stmt.lhs.pretty_print())
+
+        if isinstance(problem.stmt, Constructor):
+            # find right parameter:
+            attr = problem.attribute.end.get_name()
+            if attr not in problem.stmt.get_attributes():
+                attr_rhs = "?"
+            else:
+                attr_rhs = problem.stmt.get_attributes()[attr].pretty_print()
+            return "%s.%s = %s" % (attr_rhs, problem.attribute.get_name(), problem.stmt.pretty_print())
+
+    def get_arguments(self, problem: CompilerException) -> Dict[str, Any]:
+        return{
+            "relation": problem.attribute.get_name(),
+            "instance": problem.instance,
+            "values": problem.resultvariable.value,
+            "value": problem.value,
+            "location": problem.location,
+            "reverse": problem.reverse,
+            "reverse_example": "" if not problem.reverse else self.build_reverse_hint(problem),
+            "optional": isinstance(problem.resultvariable, OptionVariable)
+        }
+
+
+def escape_ansi(line):
+    ansi_escape = re.compile(r'(\x9B|\x1B\[)[0-?]*[ -/]*[@-~]')
+    return ansi_escape.sub('', line)
+
+
+class ExplainerFactory(object):
+
+    def get_explainers(self) -> List[Explainer]:
+        return [ModifiedAfterFreezeExplainer()]
+
+    def explain(self, problem: CompilerException) -> List[str]:
+        return [explanation for explainer in self.get_explainers() for explanation in explainer.explain(problem)]
+
+    def explain_and_format(self, problem: CompilerException) -> Optional[str]:
+        raw = self.explain(problem)
+        if not raw:
+            return None
+        else:
+            pre = """
+\033[1mException explanation
+=====================\033[0m
+"""
+            pre += "\n\n".join(raw)
+
+            if (hasattr(sys.stdout, 'isatty') and sys.stdout.isatty()) or const.ENVIRON_FORCE_TTY in os.environ:
```

Comment:
perhaps move the tty detection up one level in the call chain and just add an argument here.